active/0000-prebuild-addons.md

```diff
@@ -0,0 +1,95 @@
+
+During the build of consuming applications, if the target and environment information match one of the prebuilt asset bundles for a given addon, then those assets will be used by the build pipeline in place of the various `treeFor*` hooks that would normally execute for that addon.
+
+If a consumer wishes to build the addons from scratch with their application, they will be able to specify an `ignorePrebuiltAddons` flag in their `ember-cli-build.js`, like so:
+
+```js
+const app = new EmberApp(defaults, {
+  ignorePrebuiltAddons: true
+});
+```
+
+# How We Teach This
+
+The proposed changes here will primarily affect addon developers. Therefore, a new section will be added to the Ember-CLI ["Developing Addons and Blueprints" guides](https://ember-cli.com/extending/#developing-addons-and-blueprints) that describes how pre-building works at a high-level and how to perform it as an addon author.
+
+Application users do not really need to be aware of these changes as they will be mostly transparent from a consumption standpoint. Though, they should notice faster build times.
+
+Consumers that want to opt-out of pre-building will need to be educated about the `ignorePrebuiltAddons` option, which should be documented in the [API documentation for `EmberApp`](https://ember-cli.com/api/classes/EmberApp.html).
+
+# Drawbacks
+
+The only considerable drawbacks to these changes are increased complexity in the Ember-CLI build pipeline and slightly increased API surface area.
+
+# Alternatives
+
+The impact of not doing this is sub-optimal build performance, especially for larger applications or those using many addons.
+
+## Configurability of Targets / Environments
+
+One alternative aspect of this proposal would be to allow addon authors to customize the targets and environments for which they pre-build. While this would likely not be too difficult to implement, it risks penalties to all users due to higher installation costs with potential for minimal return.
+
+If additional, common targets/environments become present we can roll them into the preset builds. Alternatively, we can add this feature in the future, but leaving it out for now is much easier than potentially having to unwind it.
+
+## Pre-built Asset Format
+
+To limit potential download and file system I/O costs, an alternative format for the pre-built assets may be considered. Using a JSON representation of the pre-built assets would allow for storage as a single, large file which can reduce the amount of overhead in the pre-builds. This approach, however, makes understandability and debuggability of the pre-built assets much less and should likely be considered after sufficient real-world data has been collected on the current suggested approach.
+
+# Unresolved questions
+
+- Per-target installation of addons? Is it possible to publish multiple versions of addons with pre-built assets for different targets/environments? If so, this may fix the concerns over custom pre-builds.
```

Comment:
npm scoped packages maybe? where the scope is the name of the addon and then
```
// of course ember install would handle this
npm install @ember-cli-foobar/target-name
```

I don't know if Yarn has the same feature, might be a concern.

Publishing becomes more complex, but it seems it can be automated easily.